google/cloud/storage/async_writer.h

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_ASYNC_WRITER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_ASYNC_WRITER_H
+
+#include "google/cloud/storage/async_object_requests.h"
+#include "google/cloud/storage/async_token.h"
+#include "google/cloud/storage/async_writer_connection.h"
+#include "google/cloud/future.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include "absl/types/variant.h"
+#include <memory>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage_experimental {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Perform resumable uploads asynchronously.
+ *
+ * Resumable uploads allow applications to continue uploading data even after
+ * disconnects and even after application restarts. To resume an upload the
+ * library first queries the current state of the upload, the upload uses this
+ * information to send the remaining data. Applications only need to checkpoint
+ * a string, the `UploadId()`, to resume an upload even after the application
+ * itself restarts.
+ *
+ * Some data sources do not permit rewinding to an arbitrary point, for example:
+ * if the application is receiving streaming data from an external sourcein it
+ * may be impossible to shutdown the application and recover the data streamed
+ * while the application was down.
+ *
+ * Nevertheless, even with streaming data sources, resumable uploads can be used
+ * to recover from most network problems between the application and Google
+ * Cloud Storage. The application can buffer any streaming data in memory, and
+ * periodically force a flush in the resumable upload. The service will return
+ * the number of bytes persisted and the application and then discard the
+ * corresponding bytes from its buffer.
+ *
+ * Some caveats apply:
+ * - The application should be prepared for "partial flush" results, that is
```

Comment:
s/, that is/. That is, the/

Reply:
I am removing this section.  It needs member functions that will appear in a future PR.